File: lib/src/browser/authenticated_runtime.dart
import 'dart:convert';
import 'dart:io';

import '../crypto/kaalka_runtime.dart';

Map<String, dynamic> saveAuthenticatedRuntime(
  String path,
  Map<String, dynamic> session,
  String encryptionKey,
) {
  final encrypted = encryptValue(session, encryptionKey);
  File(path).writeAsStringSync(jsonEncode({'encrypted': encrypted}));
  return {'saved': true, 'path': path, 'bounded': true};
}

Map<String, dynamic> loadAuthenticatedRuntime(
    String path, String encryptionKey) {
  final raw = File(path).readAsStringSync();
  final wrapper = jsonDecode(raw) as Map<String, dynamic>;
  final enc = wrapper['encrypted'] as String;
  final decrypted = decryptValue(enc, encryptionKey);
  if (decrypted is Map<String, dynamic>) return decrypted;
  return jsonDecode(decrypted as String) as Map<String, dynamic>;
}

Map<String, dynamic> rotateAuthenticatedSession(
  String path,
  String oldKey,
  String newKey,
) {
  final session = loadAuthenticatedRuntime(path, oldKey);
  return saveAuthenticatedRuntime(path, session, newKey);
}

File: src/browser/authenticatedRuntime.ts
import { readFileSync, writeFileSync, existsSync } from "node:fs";
import { encryptValue, decryptValue } from "../crypto/kaalka.js";
import stringify from "fast-json-stable-stringify";

export type SessionState = {
  cookies?: unknown[];
  headers?: Record<string, string>;
  auth_tokens?: unknown[];
  localStorage?: Record<string, string>;
  sessionStorage?: Record<string, string>;
};

export function saveAuthenticatedRuntime(
  path: string,
  state: SessionState,
  encryptionKey: string,
): { path: string; algorithm: string; bounded: boolean } {
  const payload = encryptValue(stringify(state), encryptionKey);
  const envelope = stringify({
    algorithm: "kaalka",
    encrypted: payload.encrypted,
    version: "2.0.0",
  });
  writeFileSync(path, envelope, "utf-8");
  return { path, algorithm: "kaalka", bounded: true };
}

export function loadAuthenticatedRuntime(
  path: string,
  encryptionKey: string,
): SessionState {
  if (!existsSync(path)) {
    return { cookies: [], headers: {}, auth_tokens: [] };
  }
  const raw = JSON.parse(readFileSync(path, "utf-8")) as { encrypted: string };
  const dec = decryptValue(raw.encrypted, encryptionKey);
  return JSON.parse(dec.decrypted) as SessionState;
}

export function rotateAuthenticatedSession(
  session: SessionState,
): SessionState {
  return {
    ...session,
    headers: { ...(session.headers ?? {}), "x-kaalka-rotated": "1" },
  };
}
